File: backend/internal/models/headless.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// ==================== Headless Card Mode Models ====================

// HeadlessConversation 表示一个完整的 Headless 对话
type HeadlessConversation struct {
	gorm.Model
	SessionID       string         `gorm:"index;not null" json:"session_id"`         // HeadlessSession.ID
	ContainerID     uint           `gorm:"index;not null" json:"container_id"`       // 关联的容器 ID
	ClaudeSessionID string         `gorm:"index" json:"claude_session_id,omitempty"` // Claude 返回的 session_id（用于 --resume）
	State           string         `gorm:"default:'idle'" json:"state"`              // running | idle | error | closed
	Turns           []HeadlessTurn `gorm:"foreignKey:ConversationID" json:"turns,omitempty"`
}

// HeadlessTurn 表示一轮对话（用户输入 + Claude 响应）
type HeadlessTurn struct {
	gorm.Model
	ConversationID uint `gorm:"uniqueIndex:uniq_conv_turn,priority:1;not null" json:"conversation_id"`
	TurnIndex      int  `gorm:"uniqueIndex:uniq_conv_turn,priority:2;not null" json:"turn_index"` // 轮次序号（0, 1, 2...）

	// 用户输入
	UserPrompt   string `gorm:"type:text" json:"user_prompt"`
	PromptSource string `gorm:"default:'user'" json:"prompt_source"` // user | strategy | monitoring

	// Claude 响应（聚合后的完整响应）
	AssistantResponse string `gorm:"type:text" json:"assistant_response,omitempty"`

	// 元数据
	ModelName    string  `gorm:"column:model" json:"model,omitempty"`
	InputTokens  int     `json:"input_tokens"`
	OutputTokens int     `json:"output_tokens"`
	CostUSD      float64 `json:"cost_usd"`
	DurationMS   int64   `json:"duration_ms"`

	// 状态
	State        string     `gorm:"default:'pending'" json:"state"` // pending | running | completed | error
	ErrorMessage string     `gorm:"type:text" json:"error_message,omitempty"`
	CompletedAt  *time.Time `json:"completed_at,omitempty"`

	// 关联的原始事件（用于详细查看）
	Events []HeadlessEvent `gorm:"foreignKey:TurnID" json:"events,omitempty"`
}

// HeadlessEvent 表示单个 StreamEvent（原始事件存储）
type HeadlessEvent struct {
	gorm.Model
	TurnID     uint `gorm:"index;uniqueIndex:uniq_turn_event,priority:1;not null" json:"turn_id"`
	EventIndex int  `gorm:"uniqueIndex:uniq_turn_event,priority:2;not null" json:"event_index"` // 事件序号

	// 事件内容
	EventType    string `gorm:"not null" json:"event_type"`         // system | assistant | user | result
	EventSubtype string `json:"event_subtype,omitempty"`            // 子类型
	RawJSON      string `gorm:"type:text;not null" json:"raw_json"` // 原始 JSON
}

// HeadlessConversation 状态常量
const (
	HeadlessConversationStateRunning = "running"
	HeadlessConversationStateIdle    = "idle"
	HeadlessConversationStateError   = "error"
	HeadlessConversationStateClosed  = "closed"
)

// HeadlessTurn 状态常量
const (
	HeadlessTurnStatePending   = "pending"
	HeadlessTurnStateRunning   = "running"
	HeadlessTurnStateCompleted = "completed"
	HeadlessTurnStateError     = "error"
)

// HeadlessTurn Prompt 来源常量
const (
	HeadlessPromptSourceUser       = "user"
	HeadlessPromptSourceStrategy   = "strategy"
	HeadlessPromptSourceMonitoring = "monitoring"
)

// HeadlessEvent 类型常量
const (
	HeadlessEventTypeSystem    = "system"
	HeadlessEventTypeAssistant = "assistant"
	HeadlessEventTypeUser      = "user"
	HeadlessEventTypeResult    = "result"
)

// TableName 指定 HeadlessConversation 表名
func (HeadlessConversation) TableName() string {
	return "headless_conversations"
}

// TableName 指定 HeadlessTurn 表名
func (HeadlessTurn) TableName() string {
	return "headless_turns"
}

// TableName 指定 HeadlessEvent 表名
func (HeadlessEvent) TableName() string {
	return "headless_events"
}

File: frontend/src/types/headless.ts
// Headless Card Mode Types

// 会话状态
export type HeadlessState = 'running' | 'idle' | 'error' | 'closed';

// Stream 事件类型
export interface StreamEvent {
  type: 'system' | 'assistant' | 'user' | 'result';
  subtype?: string;
  session_id?: string;
  model?: string;
  cwd?: string;
  tools?: string[];
  message?: MessagePayload;
  result?: string;
  error?: string;
  is_error?: boolean;
  usage?: UsageInfo;
  cost_usd?: number;
  duration_ms?: number;
  is_meta?: boolean;
}

// 消息负载
export interface MessagePayload {
  content: MessageContent[];
  usage?: UsageInfo;
}

// 消息内容
export interface MessageContent {
  type: 'text' | 'thinking' | 'tool_use' | 'tool_result';
  text?: string;
  thinking?: string;
  id?: string;
  name?: string;
  input?: Record<string, unknown>;
  tool_use_id?: string;
  content?: unknown;
  is_error?: boolean;
}

// Token 使用信息
export interface UsageInfo {
  input_tokens: number;
  output_tokens: number;
  cache_creation_input_tokens?: number;
  cache_read_input_tokens?: number;
}

// 轮次信息
export interface TurnInfo {
  id: number;
  turn_index: number;
  user_prompt: string;
  prompt_source: 'user' | 'strategy' | 'monitoring';
  assistant_response?: string;
  model?: string;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number;
  duration_ms: number;
  state: 'pending' | 'running' | 'completed' | 'error';
  error_message?: string;
  created_at: string;
  completed_at?: string;
  events?: EventInfo[];
}

// 事件信息
export interface EventInfo {
  id: number;
  event_index: number;
  event_type: string;
  event_subtype?: string;
  raw_json: string;
  created_at: string;
}

// 会话信息
export interface SessionInfo {
  session_id: string;
  claude_session_id?: string;
  state: HeadlessState;
  conversation_id: number;
  current_turn_id?: number;
}

// 历史记录负载
export interface HistoryPayload {
  turns: TurnInfo[];
  has_more: boolean;
}

// 轮次完成负载
export interface TurnCompletePayload {
  turn_id: number;
  turn_index: number;
  model?: string;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number;
  duration_ms: number;
  state: string;
  error_message?: string;
}

// 错误负载
export interface ErrorPayload {
  code: string;
  message: string;
}

// 模式切换负载
export interface ModeSwitchedPayload {
  mode: 'tui' | 'headless';
  closed_sessions: number;
}

// WebSocket 请求类型
export type HeadlessRequestType =
  | 'headless_start'
  | 'headless_prompt'
  | 'headless_cancel'
  | 'load_more'
  | 'mode_switch'
  | 'ping';

// WebSocket 响应类型
export type HeadlessResponseType =
  | 'session_info'
  | 'no_session'
  | 'history'
  | 'history_more'
  | 'event'
  | 'turn_complete'
  | 'error'
  | 'mode_switched'
  | 'pty_closed'
  | 'pong';

// WebSocket 请求
export interface HeadlessRequest {
  type: HeadlessRequestType;
  payload?: Record<string, unknown>;
}

// WebSocket 响应
export interface HeadlessResponse {
  type: HeadlessResponseType;
  payload: unknown;
}

// 前端状态
export interface HeadlessSessionState {
  sessionId: string | null;
  claudeSessionId: string | null;
  state: HeadlessState;
  conversationId: number | null;
  currentTurnId: number | null;
  
  // 对话历史
  turns: TurnInfo[];
  hasMoreHistory: boolean;
  loadingHistory: boolean;
  
  // 当前轮次的实时输出
  currentTurnEvents: StreamEvent[];
  
  // WebSocket 连接状态
  connected: boolean;
  connecting: boolean;
  error: string | null;
}

// 错误代码
export const ErrorCodes = {
  INVALID_REQUEST: 'invalid_request',
  SESSION_NOT_FOUND: 'session_not_found',
  SESSION_BUSY: 'session_busy',
  PROCESS_FAILED: 'process_failed',
  MODE_CONFLICT: 'mode_conflict',
  INTERNAL_ERROR: 'internal_error',
} as const;

export type ErrorCode = typeof ErrorCodes[keyof typeof ErrorCodes];
